controller.py

```diff
@@ -426,10 +428,44 @@ def _make_request(url, on_404, on_timeout, on_exception, allow_401=False):
             )
             return INVALID_INTEGRATION_DOCUMENT.detailed(failure_detail)
 
-        library, is_new = get_one_or_create(
-            self._db, Library,
-            opds_url=opds_url,
-        )
+        library = None
+        elevated_permissions = False
+        auth_url = auth_response.final_url
+        if shared_secret:
+            # Look up a library by the provided shared secret. This
+            # will let us handle the case where the library has
+            # changed URLs (auth_url does not match
+            # library.authentication_url) but the shared secret is the
+            # same.
+            library = get_one(
+                self._db, Library,
+                shared_secret=shared_secret
+            )
+            # This gives the requestor an elevated level of permissions.
+            elevated_permissions = True
+            is_new = False
+
+        if not library:
+            # Either this is a library at a known authentication URL
+            # or it's a brand new library.
+            library, is_new = get_one_or_create(
+                self._db, Library,
+                authentication_url=auth_url
+            )
+            if opds_url:
+                library.opds_url = opds_url
+
+        if elevated_permissions and library.authentication_url != auth_url:
```

Comment:
Do you need to check the original URL as well as the final URL here for the redirect thing to work? If not, could you explain how it would work?

Reply:
This wouldn't be used in the circulation manager, because the circulation manager always sends the current address to its authentication document. But here's how it could work:

I POST to the registry with url=http://get-an-auth-document/
Registry GETs http://get-an-auth-document/ and is redirected to https://library/my-auth-document/
Registry GETs https://library/my-auth-document/ and gets an auth document.

Without this change, the `id` of the auth document is compared against http://get-an-auth-document/. With this change in place, it's compared against 
https://library/my-auth-document/.

Basically, it's more important to validate the integrity of the auth document itself (which is going into the database) than to check it against a URL that was provided as input into the registration process.

Reply:
I was thinking setting up a redirect would be an alternate way of changing your URL even if you don't have the shared secret. I'm not sure why anyone would need to post an old URL to the registry - if you can change your auth document you can probably change that too.

Reply:
I see what you're saying. There could be a follow-up to this branch that treats a redirect from a known `authentication_url` to a new `authentication_url` as a source of elevated permissions. That would let you change the URL even if you never set up a shared secret. I don't want to do that on this branch, but I'll look into doing this sooner rather than later if you've got a server that needs this feature.

Reply:
Cool, I agree.